pwiz_tools/Skyline/Model/Proteome/ProteinAssociation.cs

```diff
@@ -561,11 +610,65 @@ public void ApplyParsimonyOptions(bool groupProteins, bool findMinimalProteinLis
+            }
+        }
+
+        private Dictionary<IProteinRecord, PeptideAssociationGroup> CalculateProteinOrGeneGroups(MappingResultsInternal results, bool geneLevel, ILongWaitBroker broker)
         {
             var _peptideGroupToProteins = new Dictionary<PeptideAssociationGroup, List<IProteinRecord>>();
 
-            foreach(var kvp in AssociatedProteins)
+            var proteinOrGeneToPeptideGroup = AssociatedProteins;
+            if (geneLevel)
+            {
+                // gene to protein to peptides; the top level dictionary uses the GeneLevelEqualityComparer
+                var geneToPeptides = new Dictionary<IProteinRecord, Dictionary<IProteinRecord, List<PeptideDocNode>>>(new GeneLevelEqualityComparer());
+                foreach (var kvp in AssociatedProteins)
+                    if (!geneToPeptides.ContainsKey(kvp.Key))
+                        geneToPeptides.Add(kvp.Key, new Dictionary<IProteinRecord, List<PeptideDocNode>> { { kvp.Key, kvp.Value.Peptides } });
+                    else
+                        geneToPeptides[kvp.Key][kvp.Key] = kvp.Value.Peptides;
+                // now pick the protein with the longest sequence if it contains all the peptides, or a concatenation of all of the sequences if not
+                IProteinRecord GenerateConcatenatedSequenceIfNecessary(Dictionary<IProteinRecord, List<PeptideDocNode>> proteinToPeptides)
+                {
+                    if (proteinToPeptides.Count == 1)
+                        return proteinToPeptides.Keys.First();
+
+                    var longestProtein = proteinToPeptides.OrderByDescending(kvp2 => kvp2.Key.Sequence.Sequence.Length).First().Key;
+                    var allPeptides = proteinToPeptides.Values.SelectMany(o => o).Distinct().ToList();
+                    if (allPeptides.All(node => longestProtein.Sequence.Sequence.Contains(node.Peptide.Sequence)))
+                        return longestProtein;
+
+                    // each protein's individual metadata is kept, but all protein sequences are replaced by the concatenated sequence
+                    var concatenatedSequence = string.Join(string.Empty, proteinToPeptides.Keys.Select(p => p.Sequence.Sequence));
+                    return new FastaRecord(longestProtein.RecordIndex, 0,
+                        new FastaSequenceGroup(longestProtein.Sequence.Name,
+                            proteinToPeptides.Keys.Select(p => new FastaSequence(p.Sequence.Name,
+                                p.Sequence.Description, p.Sequence.Alternatives, concatenatedSequence)).ToList()),
+                        new ProteinGroupMetadata(proteinToPeptides.Keys.Select(p => p.Metadata).ToList()));
+                }
+                proteinOrGeneToPeptideGroup = geneToPeptides.ToDictionary(kvp => GenerateConcatenatedSequenceIfNecessary(kvp.Value),
+                    kvp => new PeptideAssociationGroup(kvp.Value.Values.SelectMany(o => o).Distinct().ToList()));

```

Comment:
You might want to make this:
`kvp => new PeptideAssociationGroup(kvp.Value.Values.SelectMany(o => o).Distinct((IEqualityComparer<PeptideDocNode>)ReferenceValue.EQUALITY_COMPARER).ToList())`
in order to avoid calling "Equals" on the PeptideDocNode's

Reply:
It seems very strange to me that you can cast Foo<object> to Foo<PeptideDocNode>.